openmmtools/storage/iodrivers.py

```diff
@@ -0,0 +1,1741 @@
+            Path to the object to assign metadata. If the object does not exist, an error is raised
+            Not passing a path in attaches the data to the top level file
+        """
+        self._check_bind_to_file()
+        path = normalize_path(path)
+        split_path = decompose_path(path)
+        if len(split_path) == 0:
+            self.ncfile.setncattr(name, value)
+        elif split_path[0].strip() == '':  # Split this into its own elif since if the first is true this will fail
+            self.ncfile.setncattr(name, value)
+        elif path in self._groups:
+            self._groups[path].setncattr(name, value)
+        elif path in self._variables:
+            self._variables[path].add_metadata(name, value)
+        else:
+            raise KeyError("Cannot assign metadata at path {} since no known object exists there! "
+                           "Try get_directory or get_variable_handler first.".format(path))
+
+    def _bind_group(self, path):
+        """
+        Bind a group to a particular path on the nc file. Note that this method creates the cascade of groups all the
+        way to the final object if it can.
+
+        Parameters
+        ----------
+        path : string
+            Absolute path to the group as it appears on the NetCDF file.
+
+        Returns
+        -------
+        group : NetCDF Group
+            The group that path points to. Can be accessed by path through the ._groups dictionary after binding
+
+        """
+        # NetCDF4 creates the cascade of groups automatically or returns the group if already present
+        # To simplify code, the cascade of groups is not stored in this class until called
+        self._check_bind_to_file()
+        path = normalize_path(path)
+        self._groups[path] = self.ncfile.createGroup(path)
+        return self._groups[path]
+
+    def sync(self):
+        if self.ncfile is not None:
+            self.ncfile.sync()
+
+    def close_down(self):
+        if self.ncfile is not None:
+            # Ensure the netcdf file closes down
+            self.sync()
+            self.ncfile.close()
+            self.ncfile = None
+
+    def _check_bind_to_file(self):
+        """
+        Bind to and create the file if it does not already exist (depending on access_mode)
+
+        """
+        if self.ncfile is None:
+            if self.access_mode is None:
+                if os.path.isfile(self.file_name):
+                    self.ncfile = nc.Dataset(self.file_name, 'a')
+                else:
+                    self.ncfile = nc.Dataset(self.file_name, 'w')
+            else:
+                self.ncfile = nc.Dataset(self.file_name, self.access_mode)
+
+    def __init__(self, file_name, access_mode=None):
+        super(NetCDFIODriver, self).__init__(file_name, access_mode=access_mode)
+        # Initialize the file bind variable. All actions involving files
+        self.ncfile = None
+        self._groups = {}
```

Comment:
So this is actually something I am going to keep around to actually make interfacing with the NetCDF file easier. The `_groups` and `_variables` objects do not just house groups and variables at the top level, they serve as a way to flatten ALL the groups and ALL the variables in the netcdf file into a single dict. 

You are also going to have the problem any time someone directly acts on the netcdf file. the idea is that the `.ncfile` variable is there if you want it/need it, but in general everything can be handled through the IODriver. I also wrote it so that you can attempt to fetch variables on the disk on the fly, so even if someone makes changes, the system can find those changes and try to work with them, just like what we have to do now where we manually have to write both Create/Resume functionality.

Specifically because that exact problem above is a pain, I made the StorageInterface to remove that.